_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,211 @@
+Because of this, having multiple versions of the same library in the classpath is generally undesireable:
+
+* Unnecessary application size increase
+* Unexpected runtime behaviour if the order of class files changes
+
+Therefore, when resolving JARs to use for compilation and packaging, most build tools will pick only one version of each library and **evict** the rest.
+
+### Source Compatibility
+Two library versions are **Source Compatible** if switching one for the other does not incur any compile errors.  
+For example, If we can upgrade `v1.0.0` of a dependency to `v1.1.0` and recompile our code without any compilation errors, `v1.1.0` is source compatible with `v1.0.0`.
+
+### Binary Compatibility
+Two library versions are **Binary Compatible** if the compiled bytecode of these versions can be interchanged without causing binary compatibility errors.  
+For example, if we can replace the class files of a library's `v1.0.0` with the class files of `v1.1.0` without any binary compatibility errors during runtime,
+`v1.1.0` is binary compatible with `v1.0.0`.
+
+**NOTE:** While breaking source compatibility normally results in binary compatibility breakages as well, they are actually orthogonal -- breaking one does not imply breaking the other.
+
+### Forwards and Backwards Compatibility
+
+There are two "directions" when we describe compatibility of a library release:
+
+**Backwards Compatible** means that a newer library version can be used in an environment where an older version is expected. When talking about binary and source compatibility,
+this is the common and implied direction.
+
+**Forwards Compatible** means that an older library can be used in an environment where a newer version is expected.  
+Forward compatibility is generally not upheld for userland libraries. It is only important in situations where an older version of a library is commonly 
+used at runtime against code that is compiled with newer version. (e.g. Scala's standard library)
+
+Let's look at an example where library `A v1.0.0` is compiled with library `C v1.1.0`.
+
+![Forwards and Backwards Compatibility]({{ site.baseurl }}/resources/images/library-author-guide/fowards_backwards_compatibility.png){: style="width: 65%; margin: auto; display: block"}
+
+`C v1.1.0 ` is **Forwards Binary Compatible** with `v1.0.0` if we can use `v1.0.0`'s JAR at runtime instead of `v1.1.0`'s JAR without any binary compatibility errors.
+
+`C v1.2.0 ` is **Backwards Binary Compatible** with `v1.1.0` if we can use `v1.2.0`'s JAR at runtime instaed of `v1.1.0`'s JAR without any binary compatibility errors..
+
+## Why binary compatibility matters
+
+Binary Compatibility matters because failing to maintain it makes life hard for everyone.
+
+* End users has to update all library versions in their whole transitive dependency tree such that they are binary compatible, otherwise binary compatibility errors will happen at runtime
```

Comment:
<s>has</s> have

Reply:
What about "have to update versions transitively in all their dependency tree such that they are binary compatible. This process is time-consuming and error prone, and it can change the semantics of end program."?